cores/esp8266/core_esp8266_wiring_pwm.cpp

```diff
@@ -59,27 +66,35 @@ extern void __analogWriteMode(uint8_t pin, int val, bool openDrain) {
   }
 
   if (analogMap & 1UL << pin) {
-  // Per the Arduino docs at https://www.arduino.cc/reference/en/language/functions/analog-io/analogwrite/
-  // val: the duty cycle: between 0 (always off) and 255 (always on).
-  // So if val = 0 we have digitalWrite(LOW), if we have val==range we have digitalWrite(HIGH)
-
     analogMap &= ~(1 << pin);
+    const bool isOpenDrain = GPC(pin) & (1 << GPCD);
+    if (isOpenDrain != openDrain) {
+      GPC(pin) ^= (1 << GPCD);
+    }
   }
   else {
-    if(openDrain) {
-      pinMode(pin, OUTPUT_OPEN_DRAIN);
-    } else {
-      pinMode(pin, OUTPUT);
-    }
+    pinMode(pin, openDrain ? OUTPUT_OPEN_DRAIN : OUTPUT);
   }
   uint32_t high = (analogPeriod * val) / analogScale;
   uint32_t low = analogPeriod - high;
   // Find the first GPIO being generated by checking GCC's find-first-set (returns 1 + the bit of the first 1 in an int32_t)
   int phaseReference = __builtin_ffs(analogMap) - 1;
+  // Per the Arduino docs at https://www.arduino.cc/reference/en/language/functions/analog-io/analogwrite/
+  // val: the duty cycle: between 0 (always off) and 255 (always on).
+  // So if val = 0 we have digitalWrite(LOW), if we have val==range we have digitalWrite(HIGH)
   if (_setPWM(pin, val, analogScale)) {
-    analogMap |= (1 << pin);
-  } else if (startWaveformClockCycles(pin, high, low, 0, phaseReference, 0, true)) {
-    analogMap |= (1 << pin);
+    if (val > 0 && val < analogScale) analogMap |= (1 << pin);
+  } else {
+    const bool detach = (val == 0 || val == analogScale);
+    // To go steady LOW or HIGH, let the waveform run into next duty cycle, if any. Then stop.
+    if (startWaveformClockCycles(pin, high, low, static_cast<uint32_t>(detach), phaseReference, 0, true)) {
+      if (detach) {
+        delay((1000 + analogFreq) / analogFreq);
+        stopWaveform(pin);
```

Comment:
Do you mean to clear analogMap(1<<pin) here on detach?

Reply:
Yes, it's stopped to permanently HIGH or LOW, so drop the bit from analogMap.